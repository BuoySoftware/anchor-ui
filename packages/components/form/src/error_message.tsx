import { WarningCircle } from "iconoir-react";
import { Box, BoxProps, Flex } from "@buoysoftware/anchor-layout";
import { Body, Markdown } from "@buoysoftware/anchor-typography";
import { useTheme } from "@buoysoftware/anchor-theme";

interface OwnProps {
  errors: string[];
  parentId: string;
}

type ErrorMessageProps = OwnProps & BoxProps;

export const ErrorMessage: React.FC<ErrorMessageProps> = ({
  errors,
  parentId,
  ...props
}): React.ReactElement => {
  const { colors } = useTheme();
  const id = `error-${parentId}`;

  return (
    <Flex
      bg="background.criticalSubdued"
      borderRadius="4px"
      data-testid={id}
      id={id}
      py="xxs"
      px="xs"
      role="alert"
      {...props}
    >
      <WarningCircle color={colors.text.critical} width={16} height={16} />
      <Box ml="xs">
        {errors.map((error, index) => (
          <Body as="div" size="m" key={`error-${index}`} color="text.critical">
            <Markdown>{error}</Markdown>
          </Body>
        ))}
      </Box>
    </Flex>
  );
};
