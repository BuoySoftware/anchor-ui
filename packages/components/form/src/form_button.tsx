import { useFormState } from "react-hook-form";
import { Button, ButtonProps } from "@buoysoftware/anchor-button";

type FormButtonProps = ButtonProps;

export const FormButton: React.FC<FormButtonProps> = ({
  children,
  ...props
}): React.ReactElement => {
  const { isSubmitting } = useFormState();

  return (
    <Button submitting={isSubmitting} {...props}>
      {children}
    </Button>
  );
};
