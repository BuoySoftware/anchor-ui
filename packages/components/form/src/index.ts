export * from "./checkbox";
export * from "./checkbox_field";
export * from "./error_message";
export * from "./form";
export * from "./form_base_errors";
export * from "./form_button";
export * from "./form_scope_provider";
export * from "./input";
export * from "./label";
export * from "./radio_field";
export * from "./select";
export * from "./select_field";
export * from "./text_area";
export * from "./text_field";
export * from "./use_conditional_field";
export * from "./use_form_field";

export { default as anchorFormsEn } from "./i18n/anchorForms.en.json";
