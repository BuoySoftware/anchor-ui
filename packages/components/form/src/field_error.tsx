import compact from "lodash/compact";
import flatten from "lodash/flatten";
import { FieldError as HookFormFieldError } from "react-hook-form";
import snakeCase from "lodash/snakeCase";
import { BoxProps } from "@buoysoftware/anchor-layout";
import { TFunction } from "i18next";
import { useTranslation } from "react-i18next";

import { useFormScope } from "./form_scope_provider";
import { ErrorMessage } from "./error_message";

export type InputType =
  | "checkbox"
  | "collectionCheckbox"
  | "hidden"
  | "phoneNumber"
  | "radio"
  | "select"
  | "text";

interface OwnProps {
  error?: HookFormFieldError;
  name?: string;
  inputId: string;
  inputType?: InputType;
}

export type FieldErrorProps = OwnProps & BoxProps;

/**
 * Look up an error message in the i18next translations. This utilizes the
 * i18next fallback mechanism for determining who to translate the error message.
 *
 * The order is as follows:
 *
 * 1. An error message in a custom provided namespace via tNamespace scoped to
 *    forms.errors and then the form scope, field name, and error typea.
 *
 * 2. An error message in the default forms namespace scoped to the form scope,
 *    field name and error type.
 *
 * 3. An error message in default forms namespace for the given input type and
 *    error type.
 *
 * 4. An error message in the anchorForms namespace for the given input type and
 *    error type.
 *
 * 5. An error message in the forms namespace for the given error type
 *
 * 6. An error message in the anchorForms namespace for the given error type
 */
function lookupErrorMessage(
  error: HookFormFieldError,
  scope: string,
  name: string,
  inputType: InputType,
  t: TFunction,
  tNamespace: string[]
): string {
  if (error.message !== undefined && error.message !== "") {
    return error.message;
  }

  const i18nName = snakeCase(name);
  const i18nErrorScope = snakeCase(scope);

  // These are translations scoped to the form's scope and field name.
  // The use of forms.errors is to look into the custom tNamespace if one was
  // provided. The second will look up in the forms default namespace.
  const fieldScopes = [
    `forms.errors.${i18nErrorScope}.${i18nName}.${error.type}`,
    `errors.${i18nErrorScope}.${i18nName}.${error.type}`,
  ];

  // Global scopes are used to look up defaults in the forms namespace or
  // the anchorForms namespace
  const globalScopes = [
    `errors.input_types.${snakeCase(inputType)}.${error.type}`,
    `errors.${error.type}`,
  ];

  const combinedGlobalScopes = tNamespace.flatMap((namespace: string) =>
    globalScopes.map((scope) => `${namespace}:${scope}`)
  );

  return t([...fieldScopes, ...combinedGlobalScopes]);
}

export const FieldError: React.FC<FieldErrorProps> = ({
  error,
  inputId,
  name,
  inputType = "text",
  ...elementProps
}): React.ReactElement | null => {
  const { scope, tNamespace = [] } = useFormScope();
  const namespace = compact(flatten([tNamespace, "forms", "anchorForms"]));
  const { t } = useTranslation(namespace);

  if (!error || !name) return null;

  const message = lookupErrorMessage(
    error,
    scope,
    name,
    inputType,
    t,
    namespace
  );

  return (
    <ErrorMessage
      errors={message.split("\n")}
      parentId={inputId}
      {...elementProps}
    />
  );
};
