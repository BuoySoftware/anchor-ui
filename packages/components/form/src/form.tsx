import kebabCase from "lodash/kebabCase";
import { useEffect, useState } from "react";
import {
  FieldValues,
  FormProvider,
  get,
  SubmitHandler,
  UseFormReturn,
} from "react-hook-form";
import { ErrorMessage } from "./error_message";
import { Box, BoxProps } from "@buoysoftware/anchor-layout";

import { FormScopeProvider, TNamespace } from "./form_scope_provider";

export type FormValidationErrors<T> = {
  [key in keyof T]: string[];
};

export type ErrorKeys<TFormValues> = Record<string, keyof TFormValues>;

interface OwnProps<TFormValues extends FieldValues> {
  "data-testid"?: string;
  id: string;
  scope: string;
  onSubmit?: SubmitHandler<TFormValues>;
  children: React.ReactNode;
  form: UseFormReturn<TFormValues>;
  errorKeys?: ErrorKeys<TFormValues>;
  mutationErrors?: Partial<FormValidationErrors<TFormValues>>;
  tNamespace?: TNamespace;
}

export type FormProps<TFormValues extends FieldValues> = OwnProps<TFormValues> &
  BoxProps;

export const Form = <TFormValues extends FieldValues>({
  "data-testid": testId,
  id,
  scope,
  children,
  errorKeys = {},
  form: { handleSubmit, ...form },
  onSubmit = () => undefined,
  mutationErrors = {},
  tNamespace,
  ...elementProps
}: FormProps<TFormValues>): React.ReactElement => {
  const [hasErrors, setHasErrors] = useState(false);

  useEffect(() => {
    if (hasErrors) {
      Object.entries(mutationErrors).forEach(([key, errors]) => {
        if (key !== "base") {
          const messages = errors?.join("\n");
          const inputKey = get(errorKeys, key, key);

          form.setError(inputKey as any, {
            type: "server",
            message: messages,
          });
        }
      });
    }
  }, [hasErrors, mutationErrors]);

  useEffect(() => {
    setHasErrors(!!Object.keys(mutationErrors).length);
  }, [mutationErrors]);

  const baseErrors = mutationErrors["base"] || [];

  return (
    <FormScopeProvider value={{ formId: id, baseErrors, scope, tNamespace }}>
      <FormProvider {...form} handleSubmit={handleSubmit}>
        <Box
          as="form"
          autoComplete="off"
          id={id}
          data-testid={testId || kebabCase(scope)}
          onSubmit={handleSubmit(onSubmit)}
          {...elementProps}
        >
          {children}
        </Box>
      </FormProvider>
    </FormScopeProvider>
  );
};
