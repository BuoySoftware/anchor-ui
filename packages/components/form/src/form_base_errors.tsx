import { useFormScope } from "./form_scope_provider";
import { ErrorMessage } from "./error_message";

export const FormBaseErrors: React.FC = (): React.ReactElement | null => {
  const { baseErrors, formId } = useFormScope();

  if (baseErrors.length === 0) return null;

  return (
    <ErrorMessage errors={baseErrors} parentId={formId} mb="form.fieldGap" />
  );
}
