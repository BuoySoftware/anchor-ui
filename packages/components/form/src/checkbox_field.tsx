import { Box, BoxProps } from "@buoysoftware/anchor-layout";
import { RegisterOptions, useFormContext } from "react-hook-form";

import { useFormField } from "./use_form_field";
import { InlineLabel } from "./inline_label";
import { Checkbox } from "./checkbox";
import { FieldError } from "./field_error";

interface OwnProps {
  name: string;
  value?: string;
  rules?: RegisterOptions;
}

type CheckboxFieldProps = OwnProps & BoxProps;

export type CheckboxFieldValue = "1" | false;

export const DEFAULT_CHECKED_VALUE: CheckboxFieldValue = "1";

export const CheckboxField: React.FC<CheckboxFieldProps> = ({
  name,
  value = DEFAULT_CHECKED_VALUE,
  rules,
  ...boxProps
}): React.ReactElement => {
  const { label, error, inputId } = useFormField({ name });
  const { register } = useFormContext();

  return (
    <Box
      as="fieldset"
      data-testid={`checkbox-field-${inputId}`}
      display="flex"
      flexDirection="column"
      mb="form.fieldGap"
      {...boxProps}
    >
      <InlineLabel
        data-testid={`label-${inputId}`}
        display="flex"
        alignItems="center"
        htmlFor={inputId}
      >
        <Checkbox
          aria-describedby={error ? `error-${name}` : undefined}
          aria-invalid={error ? "true" : "false"}
          data-testid={inputId}
          id={inputId}
          mr="xxs"
          value={value}
          {...register(name, { shouldUnregister: true, ...rules })}
        />
        {label}
      </InlineLabel>
      <FieldError
        error={error}
        inputId={inputId}
        inputType="checkbox"
        mt="s"
        name={name}
      />
    </Box>
  );
};
