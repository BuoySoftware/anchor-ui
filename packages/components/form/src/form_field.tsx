import kebabCase from "lodash/kebabCase";
import styled from "styled-components";
import { FieldError as HookFormFieldError } from "react-hook-form";
import { Box, BoxProps } from "@buoysoftware/anchor-layout";
import { Label } from "./label";

import { FieldError, InputType } from "./field_error";

interface OwnProps {
  children: React.ReactNode;
  error?: HookFormFieldError;
  inputId: string;
  inputType?: InputType;
  label: string;
  name?: string;
}

export type FormFieldProps = OwnProps & Omit<BoxProps, "errors">;

export const FormField = styled(
  ({ children, error, inputId, inputType, name, label, ...elementProps }) => {
    const testId = kebabCase(inputId);

    return (
      <Box
        data-testid={`form-field-${testId}`}
        mb="form.fieldGap"
        display="flex"
        width="input"
        flexDirection="column"
        {...elementProps}
      >
        <Label htmlFor={inputId} mb="form.labelGap">
          {label}
        </Label>
        <Box position="relative" zIndex={1}>
          {children}
        </Box>
        <FieldError
          error={error}
          inputId={inputId}
          inputType={inputType}
          mt="xxs"
          width="input"
          name={name}
        />
      </Box>
    );
  }
)<FormFieldProps>``;
