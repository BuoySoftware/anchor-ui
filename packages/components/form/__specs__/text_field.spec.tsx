import { useForm } from "react-hook-form";
import i18n from "i18next";
import { initReactI18next } from "react-i18next";

import {
  act,
  render as baseRender,
  waitFor,
  UserRenderResult,
} from "../test_utils";
import { Form, TextField } from "../src";

describe("<TextField />", () => {
  const translations = {
    errors: {
      required: "This field is required",
    },
    hints: {
      test_form: {
        my_text_field: "Optional helper text",
      },
    },
    labels: {
      my_text_field: "This is my text field",
    },
    placeholders: {
      test_form: {
        my_text_field: "Enter value",
      },
    },
  };

  i18n.use(initReactI18next).init({
    lng: "en",
    ns: ["forms"],
    resources: {
      en: {
        forms: translations,
      },
    },
  });

  const MockForm: React.FC = (): React.ReactElement => {
    const form = useForm();

    return (
      <Form
        id="test-form"
        data-testid="test-form"
        form={form}
        scope="test_form"
        onSubmit={jest.fn()}
      >
        <TextField
          data-testid="text-field-test"
          rules={{ required: true }}
          name="myTextField"
        />
        <button type="submit">Submit</button>
      </Form>
    );
  };

  function render(): UserRenderResult {
    return baseRender(<MockForm />);
  }

  const FIELD_LABEL = i18n.t("forms:labels.my_text_field");

  it("renders properly", () => {
    const form = render();

    expect(form.asFragment()).toMatchSnapshot();
  });

  context("with an error", () => {
    it("displays errors associated with the field", async () => {
      const { user, getByLabelText, getByText } = render();
      const errorMessage = i18n.t("forms:errors.required");

      await act(() => {
        user.click(getByText("Submit"));
      });

      await waitFor(() => {
        expect(getByLabelText(FIELD_LABEL)).toHaveErrorMessage(errorMessage);
      });
    });
  });

  context("with a hint", () => {
    it("displays the hint associated with the field", async () => {
      const { user, getByText } = render();
      const hint = i18n.t("forms:hints.test_form.my_text_field");

      await act(() => {
        user.click(getByText("Submit"));
      });

      expect(getByText(hint)).toBeInTheDocument();
    });
  });
});
