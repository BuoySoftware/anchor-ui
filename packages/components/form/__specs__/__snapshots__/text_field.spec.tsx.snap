// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders properly 1`] = `
<DocumentFragment>
  .c0 {
  box-sizing: border-box;
}

.c1 {
  box-sizing: border-box;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 300px;
  margin-bottom: 1.5rem;
}

.c2 {
  box-sizing: border-box;
  color: #101112;
  width: s;
  height: s;
  margin-bottom: 0.25rem;
}

.c4 {
  box-sizing: border-box;
  position: relative;
  z-index: 1;
}

.c6 {
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c8 {
  box-sizing: border-box;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c10 {
  box-sizing: border-box;
  color: #62666A;
  width: s;
  height: s;
  margin-top: 0.25rem;
}

.c7 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c3 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Segoe UI, Roboto,Helvetica Neue,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 600;
}

.c11 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Segoe UI, Roboto,Helvetica Neue,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-style: none;
  font-size: 0.75rem;
  font-weight: 400;
  line-height: 1rem;
}

.c9 {
  outline: none;
  background: none;
  border: none;
  color: inherit;
  fontsize: inherit;
  line-height: inherit;
  margin: 0;
  padding: 0;
  width: 100%;
  text-align: inherit;
  box-sizing: border-box;
}

.c5 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Segoe UI, Roboto,Helvetica Neue,sans-serif;
  display: inline-block;
  background-color: #FFFFFF;
  border: 1px solid;
  border-radius: 4px;
  color: #101112;
  font-size: 0.875rem;
  height: 40px;
  line-height: 1rem;
  margin: 0;
  padding: 0.75rem 0.75rem;
  width: 300px;
}

.c5 transition:"background 0.75s ease",
.c5 input::-webkit-input-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c5 transition:"background 0.75s ease",
.c5 input::-moz-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c5 transition:"background 0.75s ease",
.c5 input:-ms-input-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c5 transition:"background 0.75s ease",
.c5 input::placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c5:hover {
  border-color: #46494D;
}

<form
    autocomplete="off"
    class="c0"
    data-testid="test-form"
    id="test-form"
  >
    <div
      class="c1 "
      data-testid="form-field-my-text-field"
      display="flex"
      width="input"
    >
      <label
        class="c2 c3"
        color="text.primary"
        font-weight="600"
        for="my-text-field"
      >
        This is my text field
      </label>
      <div
        class="c4"
      >
        <div
          class="c0 c5"
          data-testid="input-wrapper-text-field-test"
        >
          <div
            class="c6 c7"
            display="flex"
          >
            <div
              class="c8"
            >
              <input
                aria-invalid="false"
                class="c9"
                data-testid="input-text-field-test"
                id="my-text-field"
                name="myTextField"
                placeholder="Enter value"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <p
        class="c10 c11"
        color="text.secondary"
        font-size="font-size-75"
        font-style="none"
        font-weight="400"
      >
        Optional helper text
      </p>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
